checker/cost.go

```diff
@@ -0,0 +1,430 @@
+		}
+
+		return identCost
+	}
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+
+	if sel.TestOnly {
+		return sum
+	}
+	sum = sum.Add(c.cost(sel.Operand))
+	c.exprPath[e.Id] = append(c.exprPath[sel.Operand.Id], sel.Field)
+	targetType := c.getType(sel.Operand)
+	switch kindOf(targetType) {
+	case kindMap, kindObject, kindTypeParam:
+		sum = sum.Add(selectCost)
+	}
+	return sum
+}
+
+func (c *coster) costCall(e *exprpb.Expr) CostEstimate {
+	call := e.GetCallExpr()
+	target := call.GetTarget()
+	args := call.GetArgs()
+
+	var sum CostEstimate
+
+	argTypes := make([]AstNode, len(args))
+	for i, arg := range args {
+		// TODO: && || short circuit, so min cost should only include 1st arg eval
+		// unless exhaustive evaluation is enabled
+		sum = sum.Add(c.cost(arg))
+		argTypes[i] = c.newAstNode(arg)
+	}
+
+	ref := c.checker.ReferenceMap[e.Id]
+	if ref == nil || len(ref.GetOverloadId()) == 0 {
+		return CostEstimate{}
+	}
+	var targetType AstNode
+	if target != nil {
+		if call.Target != nil {
+			sum = sum.Add(c.cost(call.Target))
+			targetType = c.newAstNode(call.Target)
+		}
+	}
+	// Pick a cost estimate range that covers all the overload cost estimation ranges
+	fnCost := CostEstimate{Min: uint64(math.MaxUint64), Max: 0}
+	for _, overload := range ref.GetOverloadId() {
+		overloadCost := c.functionCost(overload, &targetType, argTypes)
+		if overloadCost.Max > fnCost.Max {
+			fnCost.Max = overloadCost.Max
+		}
+		if overloadCost.Min < fnCost.Min {
+			fnCost.Min = overloadCost.Min
+		}
+	}
+	return sum.Add(fnCost)
+}
+
+func (c *coster) costCreateList(e *exprpb.Expr) CostEstimate {
+	create := e.GetListExpr()
+	var sum CostEstimate
+	for _, e := range create.Elements {
+		sum = sum.Add(c.cost(e))
+	}
+	return sum.Add(createListBaseCost)
+}
+
+func (c *coster) costCreateStruct(e *exprpb.Expr) CostEstimate {
+	str := e.GetStructExpr()
+	if str.MessageName != "" {
+		return c.costCreateMessage(e)
+	} else {
+		return c.costCreateMap(e)
+	}
+}
+
+func (c *coster) costCreateMap(e *exprpb.Expr) CostEstimate {
+	mapVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range mapVal.GetEntries() {
+		key := ent.GetMapKey()
+		sum = sum.Add(c.cost(key))
+
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMapBaseCost)
+}
+
+func (c *coster) costCreateMessage(e *exprpb.Expr) CostEstimate {
+	msgVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range msgVal.GetEntries() {
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMessageBaseCost)
+}
+
+func (c *coster) costComprehension(e *exprpb.Expr) CostEstimate {
+	comp := e.GetComprehensionExpr()
+	var sum CostEstimate
+	sum = sum.Add(c.cost(comp.IterRange))
+	sum = sum.Add(c.cost(comp.AccuInit))
+
+	c.env = c.env.enterScope()
+	c.env.Add(decls.NewVar(comp.AccuVar, nil)) // Don't need to track types
+	c.env = c.env.enterScope()
```

Comment:
I think you can remove the scope entry / exit based on the simplification suggestion related to simplifying identifier resolution.

Reply:
With scopes I lose 'items.@keys' field path for "k.contains" in this case:

```
input.all(k, input[k].all(k, true) && k.contains(k))
```

Reply:
I'll use a really simple stack to get around this.